cmd/immuadmin/command/hot_backup.go

```diff
@@ -0,0 +1,661 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package immuadmin
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+
+	"google.golang.org/grpc/metadata"
+
+	"github.com/schollz/progressbar/v2"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/codenotary/immudb/pkg/api/schema"
+	"github.com/codenotary/immudb/pkg/immuos"
+)
+
+const (
+	prefix            = "IMMUBACKUP"
+	latestFileVersion = 1
+)
+
+const (
+	prefixOffset     = iota
+	versionOffset    = prefixOffset + len(prefix)
+	txIdOffset       = versionOffset + 4
+	txSignSizeOffset = txIdOffset + 8
+	txSizeOffset     = txSignSizeOffset + 4
+	headerSize       = txSizeOffset + 4
+)
+
+var ErrMalformedFile = errors.New("malformed backup file")
+var ErrTxWrongOrder = errors.New("incorrect transaction order in file")
+var ErrTxNotInFile = errors.New("last known transaction not in file")
+
+type commandlineHotBck struct {
+	commandline
+}
+
+func newCommandlineHotBck(os immuos.OS) (*commandlineHotBck, error) {
+	cl := commandline{}
+	cl.config.Name = "immuadmin"
+	cl.context = context.Background()
+	cl.os = os
+
+	return &commandlineHotBck{cl}, nil
+}
+
+func (clb *commandlineHotBck) Register(rootCmd *cobra.Command) *cobra.Command {
+	clb.hotBackup(rootCmd)
+	clb.hotRestore(rootCmd)
+	return rootCmd
+}
+
+type backupParams struct {
+	output   string
+	startTx  uint64
+	append   bool
+	progress bool
+}
+
+func (cl *commandlineHotBck) hotBackup(cmd *cobra.Command) {
+	ccmd := &cobra.Command{
+		Use:   "hot-backup <db_name>",
+		Short: "Make a copy of the database without stopping",
+		Long: "Backup a database to file/stream without stopping the database engine. " +
+			"Backup can run from the beginning or starting from arbitrary transaction.",
+		PersistentPreRunE: cl.ConfigChain(cl.connect),
+		PersistentPostRun: cl.disconnect,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			file := io.Writer(os.Stdout)
+			params, err := prepareBackupParams(cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			udr, err := cl.immuClient.UseDatabase(cl.context, &schema.Database{DatabaseName: args[0]})
+			if err != nil {
+				return err
+			}
+			cl.context = metadata.NewOutgoingContext(cl.context, metadata.Pairs("authorization", udr.GetToken()))
+
+			if params.output != "-" {
+				f, err := cl.verifyBackupFile(params)
+				if err != nil {
+					return err
+				}
+				defer f.Close()
+				file = f
+			}
+
+			return cl.runHotBackup(file, params.startTx, params.progress)
+		},
+		Args: cobra.ExactArgs(1),
+	}
+	ccmd.Flags().StringP("output", "o", "-", "output file, \"-\" for stdout")
+	ccmd.Flags().Uint64("start-tx", 1, "Transaction ID to start from")
```

Comment:
it may be worth to specify an end-tx as well ? or a limit

Reply:
It is not a problem to add it, I just cannot imagine why it can be needed. I see a use case for specifying `start-tx` for incremental backup, but I'm not sure about `end-tx`

Reply:
end-tx or limit is to set some limit when the db is big enough, but might be added in the future